### ISSUE TITLE
Single instance of object is injected when using AddHttpClient

### ISSUE THREAD

**Issue by pillash** [autofac/Autofac.Extensions.DependencyInjection#48]
Autofac version: 4.9.2
Autofac.Extensions.DependencyInjection version: 4.4.0
AspNetCore version: 2.1

Repro here: https://github.com/pillash/aspnetcoreautofacissuerepro

First off, I'm not sure if this is an AutoFac issue or ASP.NET issue. When using ASP.NET Core's `AddHttpClient` to register an object for DI, I'm noticing that this object doesn't get its dependencies injected with the expected scope (in my case, per request). 

The setup: 
* `HomeController` that depends on `ILogger` and `IClient` objects
* `ILogger` that depends on an integer logger ID that is injected per request. This is registered with `InstancePerLifetimeScope`
* `IClient` that depends on `HttpClient` and `ILogger`. This is registered with ASP.NET Core's `AddHttpClient`

The issue: While `IClient` is instantiated once per request (can verify by putting breakpoint in its constructor), the `ILogger` injected into it is the same for each request. I would expect it to be a new `ILogger` per request since I registered it with `InstancePerLifetimeScope`.

Is there some other step I'm missing to have the `ILogger` injected into `Client` be the same instance that is injected into `HomeController`? Or is this expected behavior? Thanks!

**Comment by tillig** [comment]
I will likely not have time to step into this for quite a while. Here's what I'd recommend.

- Go to the Microsoft repos and find `AddHttpClient`. All of that is on GitHub. I'm on my phone right now so I can't really go research the exact location.
- Trace that out. What is it adding to the service collection? Is it adding factories or what?
- If it adds a factory, what does the factory do?

My gut tells me this adds a factory that caches a series of client objects that can be reused. If that's the case, you wouldn't be getting things from the request, you'd be getting things from the factory. The factory would be in the root container, not a request scope.

But that's just my gut. You'd have to look. This whole reverse engineering process is what we'd have to do, too - we don't have every framework we interact with memorized. Given the other priorities we have at the moment, this isn't something we'll be able to do right now.

**Comment by pillash** [comment]
Thanks, I will try to dig into this over the next week or two.

For anyone else running into this, I've worked around it for now by injecting a `IHttpClientFactory` into any classes needing an `HttpClient` instead of injecting a `HttpClient` directly. In addition to this you'll also need to change your registration from `services.AddHttpClient<IClient, Client>()` to just `services.AddHttpClient()`, and then rely solely on AutoFac's `Register...` functions to setup your `IClient`. 

From the documentation, using `IHttpClientFactory` still has the benefits of efficiently using the underlying HTTP handlers, but moves the actual creation of the `HttpClient` to the class instead of relying on the DI framework.

**Comment by alsami** [comment]
Hi @pillash!

I just cloned your repository and tried to reproduce the issue described. Sadly I don't have `netcoreapp2.1` installed so I upgraded the project and peer dependencies to 2.2.

I don't have the issue. I am getting a new logger within `HomeController` and `Client` on each request as you have defined the registration

```csharp
public void ConfigureContainer(ContainerBuilder builder)
{
    builder.Register(
        c =>
        {
            return new Logger(loggerId++);
        }).As<ILogger>().InstancePerLifetimeScope();
}
```
Here two screenshot for request 1 and request 2

![request 1](https://user-images.githubusercontent.com/2729205/62233346-b0a5d580-b3c8-11e9-95e0-94f8e5500d92.png)
![request 2](https://user-images.githubusercontent.com/2729205/62233289-8eac5300-b3c8-11e9-98d6-3d090736398b.png)

As @tillig already described, they are using a factory to create instances of `HttpClient` which are being reused for a while. I think this **might** have been a bug in the previous version implemented with 2.1, which was the initial version released for the package [Microsoft.Extensions.Http](https://www.nuget.org/packages/Microsoft.Extensions.Http/2.1.0).

So upgrading the netcoreapp version would solve the issue. I will send you a PR in a moment.
